xla/service/scatter_determinism_expander.cc

```diff
@@ -0,0 +1,456 @@
+  std::vector<HloInstruction*> scatter_updates(
+      scatter->scatter_updates().begin(), scatter->scatter_updates().end());
+  const ScatterDimensionNumbers& dim_numbers =
+      scatter->scatter_dimension_numbers();
+
+  // If the updates tensors are empty, there is no need to update the operands.
+  // The operands can be forwarded.
+  if (ShapeUtil::IsZeroElementArray(scatter_updates[0]->shape())) {
+    if (scatter_operands.size() == 1) {
+      return scatter_operands[0];
+    }
+    return scatter->parent()->AddInstruction(
+        HloInstruction::CreateTuple(scatter_operands));
+  }
+
+  // Compute the trip count for the while loop to be used for scatter. This
+  // should be the number of indices we should scatter into the operand.
+  int64_t scatter_indices_count = ScatterIndicesCount(scatter);
+  if (!IsInt32(scatter_indices_count)) {
+    // 2147483647 is the maximum value for a 32-bit signed integer (INT32_MAX).
+    return Unimplemented(
+        "Scatter operations with more than 2147483647 scatter indices are not "
+        "supported. This error occurred for %s.",
+        scatter->ToString());
+  }
+
+  // Canonicalize the scatter_indices, after which the size of its most-major
+  // dimension must be same as the while loop trip count.
+  TF_ASSIGN_OR_RETURN(scatter_indices,
+                      CanonicalizeScatterIndices(
+                          scatter_indices, dim_numbers.index_vector_dim()));
+  CHECK_EQ(scatter_indices_count, scatter_indices->shape().dimensions(0));
+
+  // Canonicalize the updates, after which the size of their most-major
+  // dimensions must be same as the while loop trip count.
+  TF_ASSIGN_OR_RETURN(scatter_updates, CanonicalizeScatterUpdates(
+                                           scatter_updates, scatter_indices,
+                                           dim_numbers, scatter_indices_count));
+
+  HloComputation* parent = scatter->parent();
+
+  // Sort the scatter indices and updates together based on the scatter indices.
+  int64_t num_indices = ShapeUtil::ElementsIn(scatter_updates[0]->shape());
+  std::vector<HloInstruction*> sorted_tensors = SortIndicesAndUpdates(
+      scatter_indices, scatter_updates, num_indices, scatter, parent);
+  HloInstruction* sorted_scalar_indices = sorted_tensors[0];
+  std::vector<HloInstruction*> sorted_updates(sorted_tensors.begin() + 1,
+                                              sorted_tensors.end());
+
+  TF_ASSIGN_OR_RETURN(std::vector<HloInstruction*> prefix_scan_updates,
+                      ComputePrefixScan(sorted_updates, sorted_scalar_indices,
+                                        scatter, parent));
+
+  HloInstruction* last_occurrence_indices = FindLastOccurrenceIndices(
+      scatter_indices, sorted_scalar_indices, scatter, parent, num_indices);
+
+  // Finally, recreate the scatter instruction with unique indices
+  auto* new_scatter = parent->AddInstruction(HloInstruction::CreateScatter(
```

Comment:
nit: directly return without assigning to new_scatter

Reply:
changed